### PR TITLE
Ship AI update part 1

### PR DESCRIPTION
**Pull request by Vivalas**
-Fixes some bugs with ship combat (mainly so that shields only start recharging cycles when they are actually damaged, same for components), 
-Adds the ability for ships to jump when heavily damaged
-Makes the number of ships in a system based on the distance to the capital of that faction.

### CODE REVIEW

File: code/controllers/subsystem/ships.dm [hunk 1]
```diff
@@ -90,6 +90,8 @@ var/datum/subsystem/ship/SSship
 		if(S.shield_strength >= initial(S.shield_strength))
 			if(S.attacking_player && S.shield_strength > starting_shields) broadcast_message("<span class=notice>Enemy ship ([S.name]) has recharged shields to 100% strength.</span>",notice_sound)
 
+	if(!find_broken_components(S))
+		S.next_repair = world.time + S.repair_time
```
Comment (monster860) [inline code/controllers/subsystem/ships.dm]: This will cause major problems

File: code/controllers/subsystem/starmap.dm [hunk 2]
```diff
@@ -209,17 +219,27 @@ var/datum/subsystem/starmap/SSstarmap
...
 
-/datum/subsystem/starmap/proc/generate_npc_ships(var/num=2)
+/datum/subsystem/starmap/proc/generate_npc_ships(var/num=0)
 	var/f_list
 	var/generating_pirates = 0
+	var/num_to_gen = 0
+
+	if(num) num_to_gen = num
+	else
+		num_to_gen = rand(current_system.danger_level - 1, current_system.danger_level + 1)
```
Comment (monster860) [inline code/controllers/subsystem/starmap.dm]: Theres literally no reason to make a new var.

File: code/controllers/subsystem/ships.dm (hunk 1 above)
Comment (Vivalas) [inline code/controllers/subsystem/ships.dm]: Why?

Comment (monster860) [inline code/controllers/subsystem/ships.dm]: It will bump the repair time every tick if there is no broken components.

Comment (Vivalas) [inline code/controllers/subsystem/ships.dm]: Which is point. It shouldn't start "repairing" until it is actually damaged.

Comment (monster860) [inline code/controllers/subsystem/ships.dm]: By the time it is,  the next repair cycle will be in like 3 minutes because it git bumped every tick.

Comment (Vivalas) [inline code/controllers/subsystem/ships.dm]: It's not cumulative. It adds repair time to world time. Repair time is always the same.

**Comment by Vivalas** [comment]
Going to close this and re-open tomorrow as a whole with the second part.

**Pull request closed without merging**